shared_ptr_debug_helper: turn assert into on_internal_error_abort

bhalevy opened the pull request:
Add `on_internal_error_abort` that prints the internal error log message and aborts unconditionally.

It allows the caller to log more information than what would have been printed just from the failed assert.

In addition, use it in the `shared_ptr_debug_helper` to replace the assert call, that is not useful in debug runs since the address sanitizer obfuscates the assert backtrace.

Fixes #1095

denesb commented on include/seastar/core/on_internal_error.hh:
```diff
@@ -54,4 +54,11 @@ bool set_abort_on_internal_error(bool do_abort) noexcept;
+/// The error will be logged to \logger and the program will be aborted,
+/// regardless of the abort_on_internal_error setting.
+/// This overload can be used to replace assert().
+void on_internal_error_abort(logger& logger, std::string_view reason) noexcept;
```
Maybe `on_internal_fatal_error()` would be a better name.

bhalevy replied:
on_fatal_internal_error sounds better IMO

denesb replied:
fine by me

bhalevy commented:
In v2 ([be0a4c6](https://github.com/scylladb/seastar/commit/be0a4c6811abf1d0dabd24e8b1fed5a9a93e3092)):
- renamed to on_fatal_internal_error
- don't fmt::format std::thread::id, just print plain text on shared_ptr_debug_helper fatal internal error.

bhalevy commented:
CI failure https://app.circleci.com/pipelines/github/bhalevy/seastar/18/workflows/bc91682a-1bec-48fe-924b-a927c11b80cd/jobs/212
is #1190 (which I'm working on separately)

xemul commented on src/core/on_internal_error.cc:
```diff
@@ -58,3 +58,8 @@ void seastar::on_internal_error_noexcept(logger& logger, std::string_view msg) n
+
+void seastar::on_fatal_internal_error(logger& logger, std::string_view msg) noexcept {
+    set_abort_on_internal_error(true);
+    on_internal_error_noexcept(logger, msg);
```
Why not just
```c++
    logger.error("{}, at: {}", msg, current_backtrace());
    abort();
```
?

bhalevy replied:
can do, I just wanted to use the common code.
maybe we can refactor the logger.error part into a common function that both places can call.

avikivity replied:
It's common code but obfuscated.

nyh commented on include/seastar/core/shared_ptr_debug_helper.hh:
```diff
@@ -60,7 +65,9 @@ public:
 private:
     void check() const {
-        assert(_cpu == std::this_thread::get_id());
+        if (__builtin_expect(_cpu != std::this_thread::get_id(), false)) {
```
I think C++20 `[[unlikely]]` is nicer, but this is also correct.

bhalevy replied:
agreed, but we still have to support c++17 in seastar

nyh reviewed (approved):
The patch itself looks good, but the sentence in the commit message said "we lose the assert backtrace in debug mode, with the address sanitizer" suggest that perhaps we'll find ourselves replacing all asserts with his new function, one by one, each time we discover that assert is missing a backtrace in debug code. Can't we fix assert. instead?

bhalevy commented:
> Can't we fix assert. instead?

I think we should, as a followup, but our experience with scylla shows that people want this `on_fatal_internal_error` helper anyway as it allows to print additional information to what would be printed with a plain assert.
Maybe I need to add that the commit message as part of the motivation for the helper.

avikivity commented on include/seastar/core/on_internal_error.hh:
```diff
@@ -54,4 +54,11 @@ bool set_abort_on_internal_error(bool do_abort) noexcept;
+/// The error will be logged to \logger and the program will be aborted,
+/// regardless of the abort_on_internal_error setting.
+/// This overload can be used to replace assert().
+void on_fatal_internal_error(logger& logger, std::string_view reason) noexcept;
```
It should be marked [[noreturn]] so the compiler knows can convert the assertion into an optimization hint.

bhalevy replied:
right. will do

bhalevy replied:
right. will do

avikivity commented on include/seastar/core/shared_ptr_debug_helper.hh:
```diff
@@ -25,9 +25,14 @@
 #include <thread>
 #include <cassert>
+#include <fmt/format.h>
+
```
Why is this include needed?

bhalevy replied:
I'm not sure. Maybe the editor added it without me noticing

bhalevy replied:
Now I remembered, my original patch printed the differing thread id:s but fmt complained they are not formattable, yet the include directive remained.

bhalevy commented:
In v2 ([b6bd5f0](https://github.com/scylladb/seastar/commit/b6bd5f01ff7fc110645711bf245fc3813c7abe3d)):
- simplified `on_fatal_internal_error`
  - improved commit message
- on_internal_error: refactor log_error_and_backtrace
- removed extraneous include of fmt/format.hh from shared_ptr_debug_helper.hh

xemul commented on src/core/on_internal_error.cc:
```diff
@@ -33,9 +33,13 @@ bool seastar::set_abort_on_internal_error(bool do_abort) noexcept {
     return abort_on_internal_error.exchange(do_abort);
 }
 
+static void log_error_and_backtrace(logger& logger, std::string_view msg) {
```
nit: `noexcept`

bhalevy replied:
added in v3
